magento/magento2#39567: Catalog price rules with SKU using "not one of" and configurable products

KrasnoshchokBohdan opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
In situation with "not one of" condition in catalog rule, \Magento\CatalogRule\Model\Rule::getMatchingProductIds works ok and result array doesn't have excluded products, but as author of issue mentioned, plugin \Magento\CatalogRuleConfigurable\Plugin\CatalogRule\Model\Rule\ConfigurableProductHandler::aroundGetMatchingProductIds adds excluded products back to result array.   

### Related Pull Requests
<!-- related pull request placeholder -->
--

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#39567

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. create a configurable product with multiple variants as simple products (or take existing one)
2. enable SKU to be used in price rules (if not set)
3. create catalog price rule, and use SKU as filter with a "not one of" option, and provide simple product SKU

Before fix:
Products including the provided SKU are affected by the price rule, because the given SKU is a sub-product of a configurable product.

After fix:
Products except the provided SKU are affected by the catalog price rule.


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

KrasnoshchokBohdan committed:
magento/magento2#39567: Catalog price rules with SKU using "not one of" and configurable products

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39567: Catalog price rules with SKU using "not one of" and configurable products

- changing Copyright

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39567

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39567

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for the contribution!

It seems the build failures are due to PR changes. Please fix those failures and also please add an automated test in accordance to [DOD](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/).

Thanks

engcom-Charlie commented:
Hi @KrasnoshchokBohdan,

As you have marked this Pull Request as draft, moving this PR to On Hold. Please feel free to to move it to Open state once you done with the implementation and the review mentioned [here](https://github.com/magento/magento2/pull/39668#pullrequestreview-2702130701).

KrasnoshchokBohdan committed:
magento/magento2#39567: Catalog price rules with SKU using "not one of" and configurable products

- changed the check criteria
- fix for unit test

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39567

KrasnoshchokBohdan commented:
https://github.com/magento run all tests

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Charlie commented:
Thank you @KrasnoshchokBohdan for working on the mentioned points. 

As the PR is ready to review moving it to Pending Review.

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

We are still seeing some failures in static test, integration test and functional test. Please fix those.

Thanks

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39567

KrasnoshchokBohdan committed:
magento/magento2#39567: Catalog price rules with SKU using "not one of" and configurable products

- new solution with child product validation
- rollback previous changes in unit test

KrasnoshchokBohdan committed:
Merge remote-tracking branch 'origin/fix-for-issue-39567' into fix-for-issue-39567

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39567: Catalog price rules with SKU using "not one of" and configurable products

- fixes for static and unit test

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39567

KrasnoshchokBohdan commented:
@engcom-Hotel Hello, I would like to know your opinion on this issue.

> Scenario: we have a Configurable product with SKU: 0001-00-group and multiple volume variants of that product as simple products with SKU like 0001-10M, 0001-25M. We would like to add a catalog price rule for every product except for some of the SKUs like 0001-25M (we want to exclude only single variants manually)
> 
> Problem: when creating a catalog price rule, when we use SKU with "not one of", and we add 0001-25M,0002-25M, the products are still affected by the price change, because the parent product (0001-00-group configurable) is matched with that rule, and expanded to every product it contains (due to code in aroundGetMatchingProductIds in ConfigurableProductHandler from Magento\CatalogRuleConfigurable\Plugin\CatalogRule\Model\Rule)

It seems I perceived this problem as - if we exclude a product using a condition (not one of ) then the rule should not apply to this product, but catalog price rule form integration test made me doubt that I understood how the solution should look like


dev/tests/integration/testsuite/Magento/CatalogRuleConfigurable/_files/configurable_product_with_percent_rule.php
```
'conditions' => [
            '1' => ['type' => Combine::class, 'aggregator' => 'all', 'value' => '1', 'new_child' => ''],
            '1--1' => ['type' => Product::class, 'attribute' => 'sku', 'operator' => '==', 'value' => 'configurable'],
        ],
```
\Magento\CatalogRuleConfigurable\Model\Indexer\Product\ProductRuleIndexerTest 
but the test expects the rule to affect child products

engcom-Hotel commented:
Hello @KrasnoshchokBohdan,

Yes you are correct, the test in the file `ProductRuleIndexerTest.php` expects the catalog price rule to affect the child products of the configurable product. 

This is evident from the `productsDataProvider` method, which includes SKUs of the child products (`simple_10` and `simple_20`) and their expected discounted prices. The tests (testExecute, testExecuteRow, testExecuteList, and testExecuteFull) validate that the catalog rule is applied to these child products by asserting their catalog rule prices.

Even when I gone through with the catalog price rule documentation [here](https://experienceleague.adobe.com/en/docs/commerce-admin/marketing/promotions/catalog-rules/price-rules-catalog-create), it seems that it is not a valid issue, the following note is written there:

```
When using the is not one of condition with a SKU product attribute and configurable product, both the parent and child product SKUs must be selected. To avoid listing all child SKUs in the rule, you can use the does not contain condition with common SKU parts of a configurable product and its child products.
```

So while creating the rule, we need to select the configurable product SKU as well, please refer to the below screenshot of catalog price rule I created:
![image](https://github.com/user-attachments/assets/93c178a3-c61a-4ec5-a997-8ca43a092535)

And also please refer to the below screencast where rules seems working fine:
[Custom.webm](https://github.com/user-attachments/assets/36b50971-5086-412e-ad2e-c2b0a5708c3d)

Please let us know if we missed anything.

Thanks

engcom-Hotel commented:
Hello @KrasnoshchokBohdan,

I think the issue for which this PR has been resolved by using the correct configuration suggested [here](https://experienceleague.adobe.com/en/docs/commerce-admin/marketing/promotions/catalog-rules/price-rules-catalog-create) on this documentation. 

Also please go through with this https://github.com/magento/magento2/pull/39668#issuecomment-2804021977 for more clarification. 

For now we are closing this PR, please let us know if you have any further inputs on this. We will be happy to reopen it and work on it.

Thanks
